### ISSUE TITLE
No component for CID <n>

### ISSUE THREAD

**Issue by ccapndave** [phoenixframework/phoenix_live_view#759]
### Environment

* Elixir version (elixir -v): 1.9.2
* Phoenix version (mix deps): 1.5.0-dev (17deab2)
* Phoenix LiveView version (mix deps): 0.10.0 (a8ec908)
* NodeJS version (node -v): 10.13.0
* Operating system: MacOS 10.14.6
* Browsers you attempted to reproduce this bug on (the more the merrier): Chrome, Firefox
* Does the problem persist after removing "assets/node_modules" and trying again? Yes/no: yes

### Behavior

I'm afraid I don't currently have a minimum case to reproduce this, but I will explain what is happening in my app in case it helps.

I have a component which contains lots of nested components.  Everything has an `id`.

1. I start on `groups/3` which displays my component with group 3 selected.
2. I click on a user, which `push_patch`es to `users/3`, which causes an assign to change.
3. The component selected user 3.
4. I click again on group 3, the URL changes to `group/3` and then I get this JS error:

```
no component for CID 6 

recursiveCIDToString | @ | phoenix_live_view.js:1
-- | -- | --
  | dynamicToBuffer | @ | phoenix_live_view.js:1
  | toOutputBuffer | @ | phoenix_live_view.js:1
  | dynamicToBuffer | @ | phoenix_live_view.js:1
  | toOutputBuffer | @ | phoenix_live_view.js:1
  | toString | @ | phoenix_live_view.js:1
  | recursiveCIDToString | @ | phoenix_live_view.js:1
  | dynamicToBuffer | @ | phoenix_live_view.js:1
  | toOutputBuffer | @ | phoenix_live_view.js:1
  | dynamicToBuffer | @ | phoenix_live_view.js:1
  | comprehensionToBuffer | @ | phoenix_live_view.js:1
  | toOutputBuffer | @ | phoenix_live_view.js:1
  | dynamicToBuffer | @ | phoenix_live_view.js:1
  | toOutputBuffer | @ | phoenix_live_view.js:1
  | toString | @ | phoenix_live_view.js:1
  | recursiveCIDToString | @ | phoenix_live_view.js:1
  | dynamicToBuffer | @ | phoenix_live_view.js:1
  | toOutputBuffer | @ | phoenix_live_view.js:1
  | toString | @ | phoenix_live_view.js:1
  | recursiveCIDToString | @ | phoenix_live_view.js:1
  | dynamicToBuffer | @ | phoenix_live_view.js:1
  | toOutputBuffer | @ | phoenix_live_view.js:1
  | dynamicToBuffer | @ | phoenix_live_view.js:1
  | toOutputBuffer | @ | phoenix_live_view.js:1
  | toString | @ | phoenix_live_view.js:1
  | value | @ | phoenix_live_view.js:1
  | eval | @ | phoenix_live_view.js:1
  | value | @ | phoenix_live_view.js:1
  | eval | @ | phoenix_live_view.js:1
  | Show 11 more frames
```

This is the array of components that is displayed:

```
0: {0: 1, s: Array(2)}
1: {0: "manageable_selector_live_component", 1: {…}, 2: "z-50", 3: "", 4: "<div data-phx-parent-id="phx-FgLjDO5dQvC0VwHB" dat…="Live.SearchLive" id="manageables-search"></div>", s: Array(6)}
2: {0: "mt-large", 1: {…}, 2: {…}, 3: "", s: Array(5)}
3: {0: "", 1: {…}, 2: {…}, 3: "", s: Array(5)}
4: {0: "", 1: "6", 2: 8, 3: "", 4: {…}, s: Array(6)}
5: {0: "Users", 1: "6", 2: 18, 3: {…}, s: Array(5)}
8: {0: "", s: Array(2)}
9: {0: "", 1: 29, 2: "", s: Array(4)}
10: {0: "", 1: 30, 2: "", s: Array(4)}
11: {0: "", 1: 31, 2: "", s: Array(4)}
12: {0: "", 1: 32, 2: "", s: Array(4)}
13: {0: "", 1: 33, 2: "", s: Array(4)}
14: {0: "", 1: 34, 2: "", s: Array(4)}
18: {0: "", s: Array(2)}
19: {0: 37, 1: "", s: Array(3)}
20: {0: 38, 1: "", s: Array(3)}
21: {0: 39, 1: "", s: Array(3)}
22: {0: 40, 1: "", s: Array(3)}
23: {0: 41, 1: "", s: Array(3)}
24: {0: 42, 1: "", s: Array(3)}
29: {0: "<a class="mod-group-icon js-visible js-navigable a…r inset-y-0 right-0 bg-center"></div>↵    ↵  </a>", s: Array(2)}
30: {0: "<a class="mod-group-icon js-visible js-navigable a…r inset-y-0 right-0 bg-center"></div>↵    ↵  </a>", s: Array(2)}
31: {0: "<a class="mod-group-icon js-visible js-navigable a…r inset-y-0 right-0 bg-center"></div>↵    ↵  </a>", s: Array(2)}
32: {0: "<a class="mod-group-icon js-visible js-navigable a…r inset-y-0 right-0 bg-center"></div>↵    ↵  </a>", s: Array(2)}
33: {0: "<a class="mod-group-icon js-visible js-navigable a…r inset-y-0 right-0 bg-center"></div>↵    ↵  </a>", s: Array(2)}
34: {0: "<a class="mod-group-icon js-visible js-navigable a…r inset-y-0 right-0 bg-center"></div>↵    ↵  </a>", s: Array(2)}
37: {0: "<a class="mod-user-icon js-visible absolute inset-…y-0 right-0 bg-center ml-auto"></div>↵    ↵  </a>", s: Array(2)}
38: {0: "<a class="mod-user-icon js-visible absolute inset-…y-0 right-0 bg-center ml-auto"></div>↵    ↵  </a>", s: Array(2)}
39: {0: "<a class="mod-user-icon js-visible absolute inset-…y-0 right-0 bg-center ml-auto"></div>↵    ↵  </a>", s: Array(2)}
40: {0: "<a class="mod-user-icon js-visible absolute inset-…y-0 right-0 bg-center ml-auto"></div>↵    ↵  </a>", s: Array(2)}
41: {0: "<a class="mod-user-icon js-visible absolute inset-…y-0 right-0 bg-center ml-auto"></div>↵    ↵  </a>", s: Array(2)}
42: {0: "<a class="mod-user-icon js-visible absolute inset-…y-0 right-0 bg-center ml-auto"></div>↵    ↵  </a>", s: Array(2)}
```

Give me a shout if there is anything I can do in order to help locate the issue.

**Comment by josevalim** [comment]
How is your phx-target written? Also, can you confirm the component you are pointing to exists in the page? You can do that by manually getting phx-target, running a query selector yourself, and see what it returns.

**Comment by ccapndave** [comment]
In this particular case I'm not using `phx-target` at all; the groups and users are `live_patch`es.  That causes `handle_params` to be called, which updates the child components.

**Comment by josevalim** [comment]
You would need to start debugging the Diff module inside LiveView. If you can isolate the issue though, I can take a look at it. :)

**Comment by ccapndave** [comment]
Alright, I shall do my best :)

**Comment by ccapndave** [comment]
So, I spent ages trying and failing to make a reproducing case, but during the course of this I discovered that if I gave the root element of my component an `id` (in the DOM), the error goes away!

In case its any more help to you, there is a `for` comprehension based on an assign which draws a `column` component, containing a `group_section` component.  When I `push_patch` to a user, the second column disappears, and then when I navigated back to the group the JS error occurs because it can't find the `group_section` inside the `column`.  Giving the `group_section` an `id` in the DOM fixes it.

Sorry I can't be more specific...
